14/umbraco-deploy/upgrades/version-specific.md

```diff
@@ -4,83 +4,108 @@ description: >-
+
+#### Async methods
+Asynchronous methods have been added to the following interfaces (in the CMS `Umbraco.Cms.Core.Deploy` namespace):
+- `IContextCache`:
+  - `GetOrCreateAsync(...)`
+- `IDataTypeConfigurationConnector`, `IImageSourceParser`, `ILocalLinkParser` and `IValueConnector`:
+  - `ToArtifactAsync(...)`
+  - `FromArtifactAsync(...)`
+- `IServiceConnector`:
+  - `GetArtifactAsync(...)`
+  - `ProcessInitAsync(...)`
+  - `ProcessAsync(...)`
+  - `ExpandRangeAsync(...)`
+  - `GetRangeAsync(...)`
+
+These methods all have a default implementation that forwards the calls to the synchronous methods (to maintain backwards compatibility). The synchronous methods have been obsoleted though and Deploy will now always call the new asynchronous methods, so implementations should be updated to start using those instead.
+
+Within Deploy, the following base classes and methods have been updated to take advantage of the asynchonous methods:
+- `ValueConnectorBase`, `RecursiveValueConnectorBase`, `DataTypeConfigurationConnectorBase` and `ServiceConnectorBase`: all synchronous methods are obsoleted and cause compiler errors when directly invoked (to avoid potential deadlocks, because they all forward to the asynchronous methods using `GetAwaiter().GetResult()`);
+- All service and value connector implementations inheriting from the above base classes have been updated to use the asynchronous methods as well;
+- `builder.DeployDataTypeConfigurationConnectors().AddCustom(...)`: both `toArtifact` and `fromArtifact` parameters now align with the `ToArtifactAsync(...)` and `FromArtifactAsync(...)` method signatures;
+- `IArtifactImportExportService.ExportArtifactsAsync(...)`: the artifacts parameter is updated to `IAsyncEnumerable<IArtifact>`, so artifacts are asynchronously created when iterating;
+- `IServiceConnector.GetArtifactsAsync(...)` and `IServiceConnectorFactory.GetArtifactsAsync(...)`: these new extension methods call `GetArtifactAsync(...)` on the relevant service connectors and returns `IAsyncEnumerable<IArtifact>`.
+
+#### Signature notification handling
+Deploy stores artifact signatures (hashes) in the database to avoid having to create the artifact to e.g. compare them between disk/database or environments. To ensure the signatures are always up-to-date, the `ISignatureService` used custom notification handling via the `RegisterHandler(...)` and `HandleExternalNotification(...)` methods (because Deploy suppresses notifications during deployment). CMS version 12 introduced the `IDistributedCacheNotificationHandler` marker interface and Deploy 12 the `IDeployRefresherNotificationHandler` that will still handle the notifications during deployments. The following methods are therefore removed:
```

Comment:
```suggestion
Deploy stores artifact signatures (hashes) in the database to avoid creating the artifact to for example compare them between disk/database or environments. To ensure the signatures are always up-to-date, the `ISignatureService` used custom notification handling via the `RegisterHandler(...)` and `HandleExternalNotification(...)` methods (because Deploy suppresses notifications during deployment). CMS version 12 introduced the `IDistributedCacheNotificationHandler` marker interface and Deploy 12 the `IDeployRefresherNotificationHandler` that will still handle the notifications during deployments. The following methods are therefore removed:
```